From e3c36eee3789b1b3db9c5e2cd4346bf3da11e021 Mon Sep 17 00:00:00 2001
From: Alexander Duyck <alexander.h.duyck@redhat.com>
Date: Tue, 28 Oct 2014 14:41:45 -0400
Subject: [net] gro: fix aggregation for skb using frag_list

Message-id: <20141028144145.4990.78259.stgit@ahduyck-vm-rhel7>
Patchwork-id: 98491
O-Subject: [RHEL7.1 net PATCH 2/2] gro: fix aggregation for skb using frag_list
Bugzilla: 1154239
RH-Acked-by: Hannes Frederic Sowa <hannes@redhat.com>
RH-Acked-by: Marcelo Leitner <mleitner@redhat.com>
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: 1154239
Brew: 8163537

Backport of upstream commit: 73d3fe6d1c6d840763ceafa9afae0aaafa18c4b5

 commit 73d3fe6d1c6d840763ceafa9afae0aaafa18c4b5
 Author: Eric Dumazet <edumazet@google.com>
 Date:   Mon Sep 29 10:34:29 2014 -0700

    gro: fix aggregation for skb using frag_list

    In commit 8a29111c7ca6 ("net: gro: allow to build full sized skb")
    I added a regression for linear skb that traditionally force GRO
    to use the frag_list fallback.

    Erez Shitrit found that at most two segments were aggregated and
    the "if (skb_gro_len(p) != pinfo->gso_size)" test was failing.

    This is because pinfo at this spot still points to the last skb in the
    chain, instead of the first one, where we find the correct gso_size
    information.

    Signed-off-by: Eric Dumazet <edumazet@google.com>
    Fixes: 8a29111c7ca6 ("net: gro: allow to build full sized skb")
    Reported-by: Erez Shitrit <erezsh@mellanox.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Alexander Duyck <alexander.h.duyck@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/core/skbuff.c b/net/core/skbuff.c
index 53afca9..2e2f598 100644
--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -3138,6 +3138,9 @@ int skb_gro_receive(struct sk_buff **head, struct sk_buff *skb)
   NAPI_GRO_CB(skb)->free = NAPI_GRO_FREE_STOLEN_HEAD;
   goto done;
  }
+ /* switch back to head shinfo */
+ pinfo = skb_shinfo(p);
+
  if (pinfo->frag_list)
   goto merge;
  if (skb_gro_len(p) != pinfo->gso_size)
-- 
1.7.1